org.eclipse.lsp4e/src/org/eclipse/lsp4e/LSPExecutor.java

```diff
@@ -0,0 +1,265 @@
+package org.eclipse.lsp4e;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+import org.eclipse.core.resources.IProject;
+import org.eclipse.jdt.annotation.NonNull;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.jface.text.IDocument;
+import org.eclipse.lsp4j.ServerCapabilities;
+import org.eclipse.lsp4j.services.LanguageServer;
+
+/**
+ *
+ */
+public abstract class LSPExecutor<E extends LSPExecutor<E>> {
+
+	/**
```

Comment:
Could all those methods be part of LanguageServers ? At the moment, LanguageServers is only containing static entry point; I imagine it could also contain those methods.

So `LanguageServers.forDocument()` would return an instance of LanguageServers, on which one could call eg `collectAll`.

Eg
```java
LanguageServers.forDocument(doc).collectAll(ls ls.completion(param))
  .thenApplyAsync(lsResponse -> toIDECompletionItems(lsResponse))
  .thenAcceptAsync(uiModel -> showInUI(uiModel), display);
```
(note that `Display` is now an Executor so it gets easier to pass it to futures)

I'm under the impression things get slightly easier if we don't have distinct LanguageServers and LSPExecutor.

Reply:
I've consolidated `LSPExecutor` onto `LanguageServers` as you suggested.

Reply:
Have followed your suggestion and moved all code into a single `LanguageServers` class.